kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/coordinator/transaction/TransactionMarkerRequestCompletionHandler.java

```diff
@@ -0,0 +1,190 @@
+                log.debug("Re-enqueuing {} transaction markers for transactional id {} under coordinator epoch {}",
+                        txnMarker.transactionResult(), transactionalId, txnMarker.coordinatorEpoch());
+            }
+
+            // re-enqueue with possible new leaders of the partitions
+            txnMarkerChannelManager.addTxnMarkersToBrokerQueue(
+                    transactionalId,
+                    txnMarker.producerId(),
+                    txnMarker.producerEpoch(),
+                    txnMarker.transactionResult(),
+                    txnMarker.coordinatorEpoch(),
+                    abortSendOrRetryPartitions.retryPartitions);
+        }
+    }
+
+    private AbortSendingRetryPartitions hasAbortSendOrRetryPartitions(
+            String transactionalId,
+            WriteTxnMarkersRequest.TxnMarkerEntry txnMarker,
+            ErrorsAndData<Optional<TransactionStateManager.CoordinatorEpochAndTxnMetadata>> errorsAndData,
+            Map<TopicPartition, Errors> errors) {
+
+        TransactionStateManager.CoordinatorEpochAndTxnMetadata epochAndMetadata = errorsAndData.getData().get();
+        TransactionMetadata txnMetadata = epochAndMetadata.getTransactionMetadata();
+
+        AbortSendingRetryPartitions abortSendingAndRetryPartitions = new AbortSendingRetryPartitions();
+
+        if (epochAndMetadata.getCoordinatorEpoch() != txnMarker.coordinatorEpoch()) {
+            // coordinator epoch has changed, just cancel it from the purgatory
+            log.info("Transaction coordinator epoch for {} has changed from {} to {}; cancel sending transaction "
+                            + "markers {} to the brokers", transactionalId, txnMarker.coordinatorEpoch(),
+                    epochAndMetadata.getCoordinatorEpoch(), txnMarker);
+            txnMarkerChannelManager.removeMarkersForTxnId(transactionalId);
+            abortSendingAndRetryPartitions.abortSending.set(true);
+        } else {
+            txnMetadata.inLock(() -> {
+                for (Map.Entry<TopicPartition, Errors> errorsEntry : errors.entrySet()) {
+                    TopicPartition topicPartition = errorsEntry.getKey();
+                    Errors error = errorsEntry.getValue();
+                    switch (error) {
+                        case NONE:
+                            txnMetadata.removePartition(topicPartition);
+                            break;
+                        case CORRUPT_MESSAGE:
+                        case MESSAGE_TOO_LARGE:
+                        case RECORD_LIST_TOO_LARGE:
+                        case INVALID_REQUIRED_ACKS: // these are all unexpected and fatal errors
+                            throw new IllegalStateException("Received fatal error " + error.exceptionName()
+                                    + " while sending txn marker for " + transactionalId);
+                        case UNKNOWN_TOPIC_OR_PARTITION:
+//                            case NOT_LEADER_OR_FOLLOWER:
```

Comment:
remove this line

Reply:
This condition was introduced in a newer Kafka client version, maybe we could retain this code and recover this code after bumping the Kafka client version, I added a comment.